File: Medii de Proiectare si Programare/TrackAndField-Java/src/main/java/repository/RegistrationRepository.java
package repository;

import model.Participant;
import model.Registration;
import model.Track;
import model.Tuple;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class RegistrationRepository implements IRegistrationRepository {
    private static final Logger logger= LogManager.getLogger();
    private JdbcUtils dbUtils;

    public RegistrationRepository(Properties props) {
        logger.info("Initializing RegistrationRepository with properties: {} ",props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Map<Participant, Integer> getParticipantsByTrackAndAgeGroup(Track track, String ageGroup) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        Map<Participant, Integer> participants = new HashMap<>();
        try(PreparedStatement preStmt = con.prepareStatement("select *, count(participantName) as [number] from Participant P " +
                "inner join Registration R on P.participantName = R.participantID " +
                "where trackID = ? and age >= ? and age <= ?")){

            preStmt.setString(1, track.getName());
            preStmt.setInt(2, Integer.parseInt(ageGroup.split("-")[0]));
            preStmt.setInt(3, Integer.parseInt(ageGroup.split("-")[1]));

            try(ResultSet result = preStmt.executeQuery()){
                while(result.next()){

                    String participantName = result.getString("participantName");
                    int age = result.getInt("age");
                    int number = result.getInt("number");

                    Participant participant = new Participant(participantName, age);

                    participants.put(participant, number);
                }
            }
        }catch (SQLException ex)
        {
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return participants;
    }

    @Override
    public Registration findOne(Tuple<Participant, Track> id) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        Registration registration = null;
        try(PreparedStatement preStmt = con.prepareStatement("select * from Participant P " +
                "inner join Registration R on P.participantName = R.participantID " +
                "inner join Track T on T.name = R.trackID " +
                "where participantID = (?) and trackID = (?)")){
            preStmt.setString(1, id.getLeft().getId());
            preStmt.setString(1, id.getRight().getId());
            try(ResultSet result = preStmt.executeQuery()){
                String participantName = result.getString("participantName");
                int age = result.getInt("age");
                String trackName = result.getString("name");
                int minimumAge = result.getInt("minimumAge");
                int maximumAge = result.getInt("maximumAge");

                Participant participant = new Participant(participantName, age);
                Track track = new Track(trackName, minimumAge, maximumAge);

                registration = new Registration(participant, track);
            }
        }catch (SQLException ex)
        {
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return registration;
    }

    @Override
    public Iterable<Registration> findAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Registration> registrations = new ArrayList<>();
        try(PreparedStatement preStmt = con.prepareStatement("select * from Participant P " +
                "inner join Registration R on P.participantName = R.participantID " +
                "inner join Track T on T.name = R.trackID")){
            try(ResultSet result = preStmt.executeQuery()){
                while(result.next()){

                    String participantName = result.getString("participantName");
                    int age = result.getInt("age");
                    String trackName = result.getString("name");
                    int minimumAge = result.getInt("minimumAge");
                    int maximumAge = result.getInt("maximumAge");

                    Participant participant = new Participant(participantName, age);
                    Track track = new Track(trackName, minimumAge, maximumAge);

                    Registration registration = new Registration(participant, track);

                    registrations.add(registration);
                }
            }
        }catch (SQLException ex)
        {
            logger.error(ex);
            System.err.println("Error DB" + ex);
        }
        logger.traceExit();
        return registrations;
    }

    @Override
    public Registration save(Registration entity) {
        logger.traceEntry("Saving registration {}", entity);
        Connection connection = dbUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into Registration(participantID, trackID) values (?, ?)")){
            preparedStatement.setString(1, entity.getParticipant().getId());
            preparedStatement.setString(2, entity.getTrack().getId());

            int result = preparedStatement.executeUpdate();

            logger.trace("Saved {} instances", result);
        }catch(SQLException ex)
        {
            logger.error(ex);
            System.err.print("Error DB" + ex);
        }
        logger.traceExit();
        return entity;
    }
}

File: Medii de Proiectare si Programare/TrackAndField-C#/TrackAndField/repository/RegistrationRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using log4net;
using TrackAndField.model;

namespace TrackAndField.repository
{
    class RegistrationRepository : IRegistrationRepository
    {
        private static readonly ILog logger = LogManager.GetLogger("RegistrationRepository");

        public RegistrationRepository()
        {
            logger.Info("Initializing RegistrationRepository class");
        }

        public IEnumerable<Registration> FindAll()
        {
            logger.InfoFormat("Finding all registrations...");

            IList<Registration> registrations = new List<Registration>();

            var connection = DBUtils.GetConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "SELECT * FROM Participant P " +
                "INNER JOIN Registration R ON P.participantName = R.participantID " +
                "INNER JOIN Track T ON T.name = R.trackID";

                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        string participantName = dataReader.GetString(0);
                        int age = dataReader.GetInt32(1);
                        string trackName = dataReader.GetString(4);
                        int minimumAge = dataReader.GetInt32(5);
                        int maximumAge = dataReader.GetInt32(6);
                        Participant participant = new Participant(participantName, age);
                        Track track = new Track(trackName, minimumAge, maximumAge);
                        Registration registration = new Registration(participant, track);
                        registrations.Add(registration);
                    }
                }
            }

            return registrations;
        }

        public Registration FindOne(Tuple<Participant, Track> id)
        {
            logger.InfoFormat("Searching registration...");

            Registration registration = null;

            var connection = DBUtils.GetConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "SELECT * FROM Participant P " +
                "INNER JOIN Registration R ON P.participantName = R.participantID " +
                "INNER JOIN Track T ON T.name = R.trackID " +
                "WHERE participantID = @p AND trackID = @t";

                var participantParameter = command.CreateParameter();
                participantParameter.ParameterName = "@p";
                participantParameter.Value = id.Item1.Name;
                command.Parameters.Add(participantParameter);

                var trackParameter = command.CreateParameter();
                trackParameter.ParameterName = "@t";
                trackParameter.Value = id.Item2.Name;
                command.Parameters.Add(trackParameter);

                using (var dataReader = command.ExecuteReader())
                {
                    if (dataReader.Read())
                    {
                        string participantName = dataReader.GetString(0);
                        int age = dataReader.GetInt32(1);
                        string trackName = dataReader.GetString(2);
                        int minimumAge = dataReader.GetInt32(3);
                        int maximumAge = dataReader.GetInt32(4);
                        Participant participant = new Participant(participantName, age);
                        Track track = new Track(trackName, minimumAge, maximumAge);
                        registration = new Registration(participant, track);
                    }
                }
            }

            return registration;
        }

        public Dictionary<Participant, int> GetParticipantsByTrackAndAgeGroup(Track track, string ageGroup)
        {
            Dictionary<Participant, int> participants = new Dictionary<Participant, int>();

            var connection = DBUtils.GetConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "SELECT *, COUNT(participantName) AS [number] FROM Participant P " +
                "INNER JOIN Registration R ON P.participantName = R.participantID " +
                "WHERE trackID = @t AND age >= @a1 AND age <= @a2";

                var trackParameter = command.CreateParameter();
                trackParameter.ParameterName = "@t";
                trackParameter.Value = track.Name;
                command.Parameters.Add(trackParameter);

                var age1Parameter = command.CreateParameter();
                age1Parameter.ParameterName = "@a1";
                age1Parameter.Value = Int32.Parse(ageGroup.Split('-')[0]);
                command.Parameters.Add(age1Parameter);

                var age2Parameter = command.CreateParameter();
                age2Parameter.ParameterName = "@a2";
                age2Parameter.Value = Int32.Parse(ageGroup.Split('-')[1]);
                command.Parameters.Add(age2Parameter);

                using (var dataReader = command.ExecuteReader())
                {
                    while (dataReader.Read())
                    {
                        string participantName = dataReader.GetString(0);
                        int age = dataReader.GetInt32(1);
                        int number = dataReader.GetInt32(4);
                        Participant participant = new Participant(participantName, age);
                        participants.Add(participant, number);
                    }
                }
            }

            return participants;
        }

        public Registration Save(Registration entity)
        {
            logger.InfoFormat("Saving registration...");

            var connection = DBUtils.GetConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "INSERT INTO Registration(participantID, trackID) VALUES (@p, @t)";

                var participantParameter = command.CreateParameter();
                participantParameter.ParameterName = "@p";
                participantParameter.Value = entity.Participant.Name;
                command.Parameters.Add(participantParameter);

                var trackParameter = command.CreateParameter();
                trackParameter.ParameterName = "@t";
                trackParameter.Value = entity.Track.Name;
                command.Parameters.Add(trackParameter);

                return entity;
            }
        }
    }
}
